APPS-1377 migrate medium integration tests to dxcint

JanProvaznik opened the pull request:
(feat) ExpectedFailureMessage, ResultsTestMixin
(feat) integration.yml - add dxcint
(refactor) ExpectedFailure

JanProvaznik committed:
dxcint

(feat) ExpectedFailureMessage, ResultsTestMixin
(feat) integration.yml - add dxcint
(refactor) ExpectedFailure

JanProvaznik committed:
dxcint

(feat) add expected failure message tests
(fix) ExpectedFailureMessage - describe execution correctly
(refactor) Messenger - minor rename

JanProvaznik committed:
dxcint

(feat) migrate all expected_failure tests
(docs) run_tests.py comment which tests were migratecd

JanProvaznik committed:
dxcint

(feat) Mixins for test customization

JanProvaznik committed:
dxcint

(feat) ExpectedOutput - test class progress (needs refactoring, testing)
(feat) AnalysisFinished
(feat) integration.yml - remove old method of testing

JanProvaznik committed:
dxcint

(refactor) ExpectedOutput
(fix) Messenger - log using Context.logger
(docs) dxcint readme - add test classes description table

JanProvaznik committed:
dxcint

(fix) ExpectedOutput
(fix) mixins shouldn't duplicate kwargs
(fix) Terraform - some mypy warnings
(feat) LockedExpectedOutput, ReorgLockedExpectedOutput test classes, "dict" (RLEO) test
(refactor) RegisteredTest - additional compiler flags and typing

JanProvaznik committed:
dxcint

(feat) add 'strings' test

JanProvaznik committed:
dxcint

(feat) migrate a lot of tests
(feat) RegisteredTestFactory - add LockedExpectedOutput
(fix) AnalysisFinished - pass when finished, fail otherwise

JanProvaznik committed:
dxcint - minor

JanProvaznik committed:
dxcint

minor

JanProvaznik committed:
dxcint

(fix) RegisteredTest - compile inputs

JanProvaznik committed:
dxcint

(feat) migrate a lot of tests (most of 'M' suite)

JanProvaznik committed:
dxcint

(fix) RegisteredTest - lock when compiling

JanProvaznik committed:
dxcint

(fix) 4 integration tests that fail because of error in expected outputs

JanProvaznik committed:
dxcint

(feat) add mixins and classes for tests with --extras
(feat) migrate tests
(fix) invalid file names in tests

JanProvaznik committed:
dxcint

(fix) StaticOnly args, test fixtures typos
(feat) migrate more tests

JanProvaznik committed:
dxcint

(feat) migrate manifest tests
(fix) ignore reuse by default
(style) reorganize imports

JanProvaznik committed:
dxcint

(feat) migrate extern tests
(feat) reintroduce old integration.yml and move new one to integration2.yml
(docs) update dxcint README with test classes explanation
(fix) move draft-2 tests to 'L' suite

JanProvaznik committed:
dxcint

(fix) RegisteredTest - remove deadlock when compiling
(fix) medium.json - name of reorg_locked_expected_output class

JanProvaznik committed:
dxcint

(fix) bugs in AppExternExpectedOutput, ExternExpectedOutput

JanProvaznik committed:
Merge branch 'develop' into APPS-1377-migrate-tests-to-dxcint

Gvaihir commented on dxcint/README.md:
```diff
@@ -80,12 +80,50 @@ results) will be copied to `dxcint/resources`.
+|test category| implementation class name | description |
+| --- | --- | --- | 
+| registered_test | RegisteredTest | always fails, parent of all tests
+| analysis_finished | AnalysisFinished | _validate checks that the analysis succeeded
```
`_validate` - code formatting with ticks

Gvaihir commented on dxcint/README.md:
```diff
@@ -80,12 +80,50 @@ results) will be copied to `dxcint/resources`.
 Then, if this category needs a new behavior when compiling, running and/or validating the test, a new subclass of 
 `RegisteredTest` needs to be implemented. There might be up to 3 methods that might need implementation.
 
+Use **Mixins** to add common functionality to your new test type.
```
TBD - is there a common place to plugin `Mixins`? Like in the `args` of test class `__init__`? If not - figure out how and where

Gvaihir replied:
EDIT: Mixins are not injected dependencies, they provide method inheritance. 
Questions to consider but not to fix right away:
Inheritance vs composition? inject as dependencies like singleton objects (i.e. classes with classmethods)?

Gvaihir commented on dxcint/dxcint/Terraform.py:
```diff
@@ -140,12 +140,13 @@ def _build_asset(self, language: str) -> str:
 
     def _create_asset_spec(self, language: str) -> Dict:
-        spec_exports = [x.export_spec() for x in self._dependencies].remove(None)
+        # spec_exports = [x.export_spec() for x in self._dependencies].remove(None) # this line doesn't do anything
```
revert: it does. It passes `exec_depends` if dependency is a package dependency

JanProvaznik replied:
The [].remove method does not have a return value, I reintroduced the correct behaviour in APPS-1432 when it was needed, I can put it here as well, it will create merge conflicts though.

Gvaihir commented on dxcint/dxcint/Terraform.py:
```diff
@@ -140,12 +140,13 @@ def _build_asset(self, language: str) -> str:
             {"name": "bzip2"},
             {"name": "jq"},
-        ] + (spec_exports or [])
+        ]
```
revert, above

Gvaihir commented on dxcint/dxcint/utils.py:
```diff
@@ -76,3 +83,120 @@ def async_retry_wrapper(*args, **kwargs):
         return async_retry_wrapper
 
     return async_retry_inner
+
```
Avoid growing `utils` - eventually they become a garbage can type module. Allocate methods to the classes when appropriate. Also could not find usage of some functions in the codebase - check if they are actually used. Delete if not

Gvaihir reviewed (changes requested):
Thanks for all the code!
Partial review:
Positives:
1. Good understanding of architecture and OOP design. 
2. Good README docs
3. Readable code and variable names

Things to work on:
1. none of the added functionality is unit-tested. Your code is only as good as it's tested. True that you have integration tests to check the functionality indirectly, but it does not validate your framework in terms of regressions. It is a lot of work upfront to mock stuff and write tests, but it will pay off. Read more about TDD in OOP, e.g. Freeman & Pryce "Growing OO Software, Guided by Tests". Absence of unit tests was the main reason previous script became hard to maintain.
2. Second thought on Mixins: would there be an alternative way to inject the functionality without multi-class inheritance? I left a thought on that in one of the comments, try to think about it before reading - maybe you arrive to the same conclusion or a better one.
3. Avoid large PRs. If it grows too big - split it in multiple smaller ones. 
4. Avoid growing the `utils` module. 
5. Docstrings - invest in documenting your code.

You don't have to change anything in this PR if you have other tasks. I will take it from here

Gvaihir committed:
Merge branch 'develop' into APPS-1377-migrate-tests-to-dxcint

JanProvaznik committed:
dxcint (#394)

(test) add unit tests (and declarations of needed tests) for testclasses and mixins
(docs) clarify how to use mixins
(fix) Terraform.py: reintroduce spec_exports (same as in APPS-1432)
(refactor) move describe api calls to Messenger

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(feat) README.md - clarifications

Gvaihir committed:
Merge branch 'develop' into APPS-1377-migrate-tests-to-dxcint

# Conflicts:
#	scripts/run_tests.py

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) Terraform.py - remove None in place in the separate line

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) conftest.py - fixture scope
(fix) elsewhere - documentation

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) conftest.py - fixture scopes

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) everywhere - refactoring, version updates

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) everywhere - reverting

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) everywhere - refactoring

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) everywhere - reverting

Gvaihir committed:
APPS-1377 Migrate medium test set to dxcint
(fix) everywhere - reverting

Gvaihir committed:
Merge branch 'develop' into APPS-1377-migrate-tests-to-dxcint

# Conflicts:
#	scripts/run_tests.py

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) conftest.py - prerequisites
(fix) elsewhere - minor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) Messenger.py - describe property
(fix) elsewhere - minor

Gvaihir committed:
APPS-1341 dxni frag inputs
(feat) TestDiscovery.py - json indent

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) Messenger.py - describe property
(fix) elsewhere - moving pytest fixtures, refactor, adding unit tests

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) test_ExpectedOutput.py - init

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) everywhere - linter

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) scala.yml - don't run dxC unit tests if dxcint label is present

Gvaihir committed:
Merge branch 'develop' into APPS-1377-migrate-tests-to-dxcint

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) conftest.py - added fixture

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) test_ExpectedOutput.py - tests and fixtures

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) test_ExpectedOutput.py - tests
(fix) elsewhere - refactoring

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) ExtrasMixin.py - refactor + test

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) LockedMixin.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) ManifestMixin.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) ReorgMixin.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) StaticOnlyMixin.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) PinnedInstanceMixin.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) ExpectedFailureMessage.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) ExpectedFailureMessage.py - tests
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) dxcint.yml - install

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) dxcint.yml - install version

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) test_LockedExpectedOutput.py - test
(fix) ecr_docker_extras.json - updated fixture

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) array_structs* - moved to correct category

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) dxcint.yml - install dxpy

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) dxcint.yml - install dxpy

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) dxcint.yml - install dxpy

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) dxcint.yml - install dxpy

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) dxcint.yml - install dxpy

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) elsewhere - refactor

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) ExternExpectedOutput.py - tests

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) ExternExpectedOutput.py - tests

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(fix) AppExternExpectedOutput.py - tests

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) run_tests.py - comments on what's migrated

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) mock_1_input.dx.json - test fixture

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA

Gvaihir committed:
APPS-1377 migrate moderate tests to dxcint
(feat) integration2.yml - updated integration tests GHA
